### ISSUE TITLE
Extra encoder for 

### ISSUE THREAD

**Issue by Arrow7000** [thoth-org/Thoth.Json#169]
I have the following type that I want to encode:

```fsharp
type CstNode<'a> =
    { node : 'a
      source : TokenWithSource list }
```
but I only want to encode the `node` field's value directly, because the `source` is not necessary for serialisation, and it would just add clutter to the generated JSON.

I then try to encode a sample record using the following code

```fsharp
let cstNodeEncoder : Encoder<CstNode<'a>> =
    fun (cstNode : CstNode<'a>) ->
        printfn "Calling CstNode encoder" // Log this encoder
        // Encode only the .node field using an auto encoder
        Encode.Auto.generateEncoder () cstNode.node

// We don't care about the decoder because we are only ever going to encode the type
let cstNodeDecoder : Decoder<CstNode<'a>> = fun _ -> failwithf "Not implemented"

let cstCoder =
    Extra.empty
    |> Extra.withCustom cstNodeEncoder cstNodeDecoder

let toJson result =
    Encode.Auto.toString (4, value = result, extra = cstCoder)

[<EntryPoint>]
let main argv =
    { node = 1234; source = List.empty }
    |> List.singleton
    |> toJson
    |> printfn "%A"

    0
```

Yet I still consistently get this in my console

```
"[
    {
        "node": 1234,
        "source": []
    }
]"
```

So it looks like either `Encode.Auto.generateEncoder () cstNode.node` isn't actually doing anything, or that my extra encoder is never even being called. I'm leaning towards the latter because the `printfn` statement in the body of the encoder is never appearing in my console, so it looks like it's never being called. 

Is it possible that this is because my type includes a generic type parameter which doesn't play nicely with the logic for checking whether a custom encoder needs to be invoked?

**Comment by MangelMaxime** [comment]
I think the problem is coming from how we identify custom coders:

https://github.com/thoth-org/Thoth.Json/blob/ca5930a39158f76f2ede7ef1b44668c033684775/src/Extra.fs#L11-L16

We are using `typeof<'Value>.FullName` which contains the generic information `
Microsoft.FSharp.Collections.FSharpList`1[System.Int32]`.

Supporting your case will require to find a way to identify types with generic not filled and also change some mechanism in how we resolve the auto encoder and decoder. I don't know if detecting types with generic not provided is doable or not

**Comment by Arrow7000** [comment]
What about instead of always getting the full name of the provided type exactly as is, if we get the generic version of the type, both when inserting into the map and when retrieving it? Something like this:
```fsharp
let inline internal withCustomAndKey (encoder: Encoder<'Value>) (decoder: Decoder<'Value>)
           (extra: ExtraCoders): ExtraCoders =
    let theType = typeof<'Value>

    // Get the type with any concrete generic parameters reset so that no matter what the specific type parameters are set, this returns the generic type
    let genericVersionOfType = theType.GetGenericTypeDefinition ()

    { extra with
          Hash = System.Guid.NewGuid().ToString()
          Coders =
              extra.Coders |> Map.add genericVersionOfType.FullName (Encode.boxEncoder encoder, Decode.boxDecoder decoder) }
```

and similarly at the point where we look for extra coders, instead of https://github.com/thoth-org/Thoth.Json/blob/ca5930a39158f76f2ede7ef1b44668c033684775/src/Encode.fs#L441-L443

we could do something like:
```fsharp
and private autoEncoder (extra: Map<string, ref<BoxedEncoder>>) caseStrategy (skipNullField : bool) (t: System.Type) : BoxedEncoder =
      let genericVersionOfType = t.GetGenericTypeDefinition ()
      let fullname = genericVersionOfType.FullName
      match Map.tryFind fullname extra with
      | Some encoderRef -> fun v -> encoderRef.contents v
```

**Comment by Arrow7000** [comment]
One question that arises with the above is whether we want to include logic for supporting different coders for the same type if it has different type parameters set. 

I.e. do we want to let the user define different encoders for `CstNode<'a>`, `CstNode<int>`, and `CstNode<string>`, with the latter two being chosen for `int`s and `string`s respectively, and the former being chosen if it's any other type.

Personally I don't think this is necessary because if the user really wants to encode/decode different things differently they should just create a discriminated union containing different types in each variant and have different encoders for the different types in each variant.

So I think the initial idea of using the type-parameter-less name of the type is all that's needed to make this work.

**Comment by MangelMaxime** [comment]
`GetGenericTypeDefinition` seems to be support by Fable so I should be able to make the changes in Thoth.Json.

I will see if this breaks any tests or not.



> One question that arises with the above is whether we want to include logic for supporting different coders for the same type if it has different type parameters set.

I think that if the user want different representation depending on the "context" then he should use DUs or manual coder to have full control over what he does.

In general, I think that a type as a single way of being represented. If this is not the case, user should use manual coders.

**Comment by Arrow7000** [comment]
Yes, agreed. Thank you. I can have a crack at it myself also if that would be helpful?

**Comment by Arrow7000** [comment]
@MangelMaxime I've had a go at making the changes myself. I haven't been able to run the tests because I can't get things to build on VS, but I've pushed up my work to this branch https://github.com/thoth-org/Thoth.Json.Net/compare/main...Arrow7000:Thoth.Json.Net:use-generic-names-in-extras-map

Lmk if you want me to make more changes and/or submit a PR 🙂

**Comment by MangelMaxime** [comment]
Hello @Arrow7000 thank you for your work.

I didn't forget about your issue. I am trying to wrapping up another work that I already have started and will have a look at your issue and code after.

I see that you made the changes for Thoth.Json.Net, could you open a PR on that repo and also make the same changes for Thoth.Json? Both projects being still splited we need to do the job twice... 😅

**Comment by MangelMaxime** [comment]
The new API from Thoth.Json.Core.Auto now works with the original issue reported.
